### PR TITLE
Merge master into feature/py3

### PR DESCRIPTION
**Pull request by bernhardkaindl**
# Merge master into feature/py3
@stephenchengCloud - of course this has to be approved by you
## Motivation

@psafont said that master should be merged into `feature/py3` before applying #5622 (https://github.com/xapi-project/xen-api/pull/5622#issuecomment-2104249026):
> I would prefer if master was merged to feature/py3, then add the necessary patches on top (like the pytype and pyright fixes)

@stephenchengCloud, @psafont wrote in https://github.com/xapi-project/xen-api/pull/5622#issuecomment-2104358767:
> My preference would be the responsibility to merge these changes to be transferred over to stephen and other working on the python3 feature so they can prioritise this to and use these contributions easier as they see fit.

### I gladly accept:
- I have no further tasks to submit.
- My work on it is done.
- I had some time left, so I thought I could submit the missing merge of master to feature/py3.

## Manually resolved merge-conflicts

There were only a few areas that needed manual work to resolve the merge:
* `pyproject.toml`:
  - `feature/py3` updated the `pylint` configuration (e.g. suppressed `pylint` messages) etc
  - master as well, and this caused also a duplicate config section, largely equivalent
  - resolved using changes from `master` which were more nicely formatted and documented for those.
  - master removed issuing python warnings from the `scripts` directory to reduce noise in PRs.
  - for those, keep the warnings to keep the remaining Python2 work visible.
* `.github/workflows/main.yml` -> * `.github/workflows/other.yml`
  - `master` moved the non-OCaml parts to other.yml, while some changes were made in py3.
  - Merged the needed changes into the new `.github/workflows/other.yml`
* In `feature/py3`, a few tests and files were moved and created
  - With the merge of the configuration updates in pyproject.toml, as 
  
### List of all resolved merge-conflict and suppressions of warnings
@stephenchengCloud:
The files that needed manual merging and suppression of warnings are:
- `pyproject.toml`
- `.github/workflows/main.yml`
- `.github/workflows/other.yml`
- `python3/tests/import_helper.py`
- `python3/libexec/usb_scan.py`
- `scripts/mail-alarm`
- `scripts/examples/python/provision.py`

I added these suppressions in this PR to fix CI with the merge of `pyproject.toml` with master.

### CODE REVIEW

**Review by psafont — APPROVED** [review]

File: .github/workflows/other.yml [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+          --cov-report xml:.git/coverage${{matrix.python-version}}.xml
+        env:
+          PYTHONDEVMODE: yes
+
+      - name: Run Pytest for python 3 and get code coverage for Codecov
+        if: ${{ matrix.python-version != '2.7' }}
+        run: >
+          pytest
+          --cov=scripts --cov=ocaml/xcp-rrdd --cov=python3/
+          scripts/ ocaml/xcp-rrdd python3/ -vv -rA
```
Comment (stephenchengCloud) [inline .github/workflows/other.yml:76]: line 75 and line76:
These two lines are modified in feature branch to be:
```
--cov=python3/
python3/tests -vv -rA
```

File: .github/workflows/other.yml [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        env:
+          SKIP: no-commit-to-branch
+
+      - name: Install dependencies only needed for python 2
+        if: ${{ matrix.python-version == '2.7' }}
+        run: pip install enum
+
+      - name: Install dependencies only needed for python 3
+        if: ${{ matrix.python-version != '2.7' }}
+        run: pip install opentelemetry-api opentelemetry-exporter-zipkin-json opentelemetry-sdk pandas pytype toml wrapt pyudev
```
Comment (stephenchengCloud) [inline .github/workflows/other.yml:54]: line 54:
This line has been modified in the feature branch to be:
`run: pip install pandas pytype toml wrapt`

File: .github/workflows/other.yml [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+      - name: Install common dependencies for Python ${{matrix.python-version}}
+        run: pip install future mock pytest-coverage pytest-mock
+
+      - name: Run Pytest for python 2 and get code coverage for Codecov
+        if: ${{ matrix.python-version == '2.7' }}
+        run: >
+          pytest
+          --cov=scripts --cov=ocaml/xcp-rrdd
```
Comment (stephenchengCloud) [inline .github/workflows/other.yml:63]: line 63:
This line has been modified in the feature branch to be:
`--cov=ocaml/xcp-rrdd`

**Comment by stephenchengCloud** [comment]
Because the Python CI config was moved to `other.yml` from `main.yml`  in the master branch recently, `git merge` cannot detect the conflicts in this newly added  `other.yml` file. Need to manually move the changes in `main.yml` made by the feature/py3 branch to `other.yml`.
Check my comments above. Not sure if I missed anything. Really thank @bernhardkaindl for this, it's time-consuming work.

File: .github/workflows/other.yml (hunk 1 above)
Comment (bernhardkaindl) [reply .github/workflows/other.yml:76]: > line 75 and line76:
> These two lines have been modified in the feature branch to be:
> ```
> --cov=python3/
> python3/tests -vv -rA
> ```
That may be, but `ocaml/xcp-rrdd` contains a Python script, that while not being made incompatible with Python2, is already called using Python3, I'm understood.

Therefore, we should consider `ocaml/xcp-rrdd` to be a Python3 directory and should test and cover it.

If `scripts/` is tested or not in, this line is a nitpick and I can check if I can remove it.

But to get the merge done quickly, without a lot of back and forth, Ashwin can remove it also after the merge if needed.

File: .github/workflows/other.yml (hunk 2 above)
Comment (bernhardkaindl) [reply .github/workflows/other.yml:54]: Yes, but more tests have been merged in `master`.
Thus, this is a merge from `master`.

We need those `opentelemetry-sdk` packages for the new tests for observer.py from `master`.

The master branch here merges and takes precedence in order to run CI.

File: .github/workflows/other.yml (hunk 3 above)
Comment (bernhardkaindl) [reply .github/workflows/other.yml:63]: That may be, but as long as there are Python scripts in the `scripts` directory, we should not remove testing them.

**Comment by bernhardkaindl** [comment]
> Because the Python CI config was moved to other.yml from main.yml in the master branch recently, git merge cannot detect the conflicts in this newly added other.yml file. Need to manually move the changes in main.yml made by the feature/py3 branch to other.yml.
> Check my comments above. Not sure if I missed anything. Really thank @bernhardkaindl for this, it's time-consuming work.

Hi @stephenchengCloud, I answered your comments and I don't think that there is a problem with the changes as is, and we can merge this as is.

Finer adjustments can always be made after the merge.

The main point is that we should not be picky and make it more time-consuming by wanting this to be "perfect":

Instead, we can do minor adjustments after the merge.

**Review by bernhardkaindl — COMMENTED** [review]
> Because the Python CI config was moved to other.yml from main.yml in the master branch recently, git merge cannot detect the conflicts in this newly added other.yml file. Need to manually move the changes in main.yml made by the feature/py3 branch to other.yml.
> Check my comments above. Not sure if I missed anything. Really thank @bernhardkaindl for this, it's time-consuming work.

Hi @stephenchengCloud, I reviewed and answered your comments above:

The locations you commented on are indeed correct as they are now.

This PR is correct to merge as-is now.

**Review by stephenchengCloud — APPROVED** [review]

**Review by liulinC — APPROVED** [review]

**Comment by liulinC** [comment]
Merge this PR for now, and fix the comments/issues in future branch.

**Pull request merged**